### DISCUSSION TITLE
Add Self Hosted configuration out of azure

### DISCUSSION THREAD

**Discussion by felpasl** [community/community#160665] [Actions]
**Tool information**
- Tool name: Self Hosted agent configuration
- Add or update? add
- Desired version: 
- Approximate size: 
- If this is an add request:
  - Brief description of tool: create a sh option to configure any environment to act as a self build host, in other clouds or in on-premisses vm
  - URL for tool's homepage:

**Area for Triage**:

**Question, Bug, or Feature?**: Feature

**Virtual environments affected**
- [ ] macOS 10.15
- [ ] Ubuntu 16.04 LTS
- [ ] Ubuntu 18.04 LTS
- [ ] Windows Server 2016 R2
- [ ] Windows Server 2019

**Can this tool be installed during the build?**

**Are you willing to submit a PR?**

**Comment by oleg-brizy** [comment]
[Similar](https://github.com/actions/virtual-environments/issues/666)

**Comment by alepauly** [comment]
Hi @felpasl. I think it would be interesting to have the json files be able to generate images in other packer builders such as AWS, HyperV, Virtual Box or whatever else is useful. The team that maintains this repo won't be able to take on that type of work, but we would be most welcoming of community contributions that can be shared by all around here. We'll be happy to review and merge as long as the changes don't introduce maintenance burden on the generation of the virtual-environments for GitHub Actions. I'll leave this issue open and label it as `community-feature`. I'm thinking we can use that label to denote those features that the team behind virtual-environments won't be able to get done due to priorities on the product we serve but that could be useful for the community to build and support.

**Comment by felpasl** [comment]
Sorry for my late response, yes you understand, is to use Paker to generate images out of azure and remain these files in the repository, change the builders on file, I'm not sure if it is only that.

**Comment by pekhota** [comment]
I did this for aws, but it seems that you have some additional provision script which is needed to be run after instance is ready. Can you share some ideas here. Thank you.

**Comment by oliveirafilipe** [comment]
> I did this for aws, but it seems that you have some additional provision script which is needed to be run after instance is ready. Can you share some ideas here. Thank you.

@pekhota You said that you did this for aws, I assume you did a Packer file to create an AMI. Is that available somewhere?

**Comment by oliveirafilipe** [comment]
I made a few modifications to build an AWS AMI (ubuntu 18 and 20), nothing related to limits/OS configuration, but I still cannot get the same behavior as the Github Hosted Virtual Enviornments :( Even using a [m5.large](https://aws.amazon.com/ec2/instance-types/m5/#Product_details).

In my case, I run tests with jest. With the Github Hosted Virtual Envs. the tests completes successfully, but in my self runner they get stuck, no errors.

I'll keep trying, maybe it could be something related to the agent installation, when the VM is up and running.

**Comment by jojand** [comment]
I have the following proposal to be able to build the AWS AMI image to allow running EC2 self hosted runner: [[ubuntu] add aws ami source](https://github.com/wandera/runner-images/pull/1) (the pr is not to upstream yet).

I needed to [add a new source](https://github.com/wandera/runner-images/pull/1/files#r1425002439). Also I needed to make few modifications to be able to run the packer on clean canonical base image `ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-*` [change in configure-environment.sh](https://github.com/wandera/runner-images/pull/1/files#r1425000410) and [skip waagent run](https://github.com/wandera/runner-images/pull/1/files#r1425001487) for non-azure sources.

For the backward compatibility purposes the [sources](https://github.com/wandera/runner-images/pull/1/files#r1425003812) value is taken from [variable with default value](https://github.com/wandera/runner-images/pull/1/files#r1425004250).

Does this align with your long term vision for this repo and would you be willing to accept this (possibly after some modifications) to upstream?

**Reply by asegura2026** [reply]
pls add windows support too

**Reply by j0nathontayl0r** [reply]
@jojand, before finding this question and subsequent PR, I spent some time looking into this. 

I forked the repo so haven't got something ready to provide yet, but from what I saw and read, IMO it makes more sense to make a new Packer provision as a seperate Packer file.

I created file "ubuntu-22.04-aws.pkr.hcl" under:

$repoRoot/images/ubuntu/templates/ubuntu-22.04-aws.pkr.hcl

My build (haven't tested deployment in CloudFormation yet) but the build is already working in AWS and it required minimal changes compared to the original Ubuntu 22 Azure packer build file. Did you consider this approach first? Did you find any gotchas that I'm not aware of?

**Reply by jojand** [reply]
@j0nathontayl0r creating a separate `ubuntu-22.04-aws.pkr.hcl` means you must duplicate the code that defines the provisioners in `ubuntu-20.04.pkr.hcl`. I wanted to avoid this.

**Reply by jojand** [reply]
@xxsumsxx unfortunately, supporting windows is not in our focus. In our team we just need to add AWS EC2 support.
